protocols/v2/roles-logic-sv2/src/selectors.rs

```diff
@@ -65,14 +65,56 @@ impl<Down: IsMiningDownstream> DownstreamMiningSelector<Down>
     }
 
     fn remove_downstreams_in_channel(&mut self, channel_id: u32) -> Vec<Arc<Mutex<Down>>> {
-        self.channel_id_to_downstreams
+        let downs = self.channel_id_to_downstreams
             .remove(&channel_id)
-            .unwrap_or_default()
+            .unwrap_or_default();
+        for d in &downs {
+            self._remove_downstream(d.clone());
+        };
+        downs
     }
 
+    fn _remove_downstream(&mut self, d: Arc<Mutex<Down>>) {
+        let mut ks = vec![];
+        for (k,v) in &self.request_id_to_remotes {
+            if Arc::ptr_eq(v, &d) {
+                ks.push(k.clone())
+            }
+        };
+        for k in ks {
+            self.request_id_to_remotes.remove(&k).unwrap();
+        };
+
+
+        let mut ks = vec![];
+        for (k,v) in &self.channel_id_to_downstream {
+            if Arc::ptr_eq(v, &d) {
+                ks.push(k.clone())
+            }
+        };
+        for k in ks {
+            self.channel_id_to_downstream.remove(&k).unwrap();
+        };
+    }
+
+    fn remove_downstream(&mut self, d_: Arc<Mutex<Down>>) {
```

Comment:
This and `_remove_downstream` look like they can both accept `&Arc<Mutex<Down>>` instead of `Arc<Mutex<Down>>` (results in not having to increase/decrease the atomic reference count unnecessarily).